canonical/netplan
# Ignore bad NetDefs and files via parser flags

## Description

This is an implementation of what I called "parser flags". The idea is to enable the user to change some parsing decisions.

The main application at the moment is to support ignoring parsing errors. Currently Netplan will not generate any configuration if a little mistake is made in one of its YAML files. It can be really bad if you're operating a remote system and end up rebooting the system with a syntax issue in one of your files. Your system will basically boot without any network configuration.

The IGNORE_ERRORS flag will allow the Netplan generator to ignore bad files and bad netdefs so it will still generate some configuration.

**TESTS**

Integration tests will be executed twice now, once without the ignore_errors flags and once with it enabled. It's intended to increase the confidence that the parser behaves as expected when the flag is enabled and the configuration is good.

I prepared a PPA for Ubuntu Noble with this patch: https://launchpad.net/~danilogondolfo/+archive/ubuntu/netplan-parser-flags

Suggestions of tests:

It can be easily tested in a LXD VM. You can add some broken configuration and reboot the VM or call `/usr/libexec/netplan/generate -i` to see the parsing process when it's ignoring errors

1. Break the default `50-cloud-init.yaml` and reboot the VM. For example, the config below should still bring enp5s0 up and with DHCP working.

```yaml
network:
  version: 2
  ethernets:
    enp5s0:
      dhcp4: true
      dhcp: true
```

2. Add a second file that is not a valid YAML
3.  Add a third broken file with some complex configuration and dependencies between netdefs. For example, with the configuration below `eth2` should still have a backend configuration file and it should contain `Bond=bond0`.
```yaml
network:
  ethernets:
    eth0: {}
    eth1: {}
    eth2:
      a: b

  bonds:
    bond0:
      interfaces:
        - eth0
        - eth1
        - eth2
```
4. Add a fourth broken file with missing dependencies. For example, with the config below you still should have a `br0` interface created:
```yaml
network:
  bridges:
    br0:
      addresses:
        - 10.20.30.40/24

      interfaces:
        - aaaa
        - bbbb
        - dddd
```

## Checklist

- [ ] Runs `make check` successfully.
- [ ] Retains 100% code coverage (`make check-coverage`).
- [ ] New/changed keys in YAML format are documented.
- [ ] \(Optional\) Adds example YAML for new feature.
- [ ] \(Optional\) Closes an open bug in Launchpad.
